applications/MeshingApplication/external_includes/trigen_glass_forming.h

```diff
@@ -0,0 +1,1363 @@
+    boost::numeric::ublas::bounded_matrix<double,2,2> mJinv; //inverse jacobian
+    array_1d<double,2> mC; //center pos
+    array_1d<double,2> mRhs; //center pos
+    //NodeEraseProcess* mpNodeEraseProcess;
+
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+    void RemoveCloseNodes(ModelPart& ThisModelPart, KdtreeType& nodes_tree1, NodeEraseProcess& node_erase, double& h_factor)
+    {
+        //unsigned int bucket_size = 20;
+
+        //performing the interpolation - all of the nodes in this list will be preserved
+        unsigned int max_results = 100;
+        //PointerVector<PointType> res(max_results);
+        //NodeIterator res(max_results);
+        PointVector res(max_results);
+        DistanceVector res_distances(max_results);
+        Node<3> work_point(0,0.0,0.0,0.0);
+
+        unsigned int n_points_in_radius;
+        //radius means the distance, closer than which no node shall be allowd. if closer -> mark for erasing
+        double radius;
+
+        for(ModelPart::NodesContainerType::const_iterator in = ThisModelPart.NodesBegin(); in != ThisModelPart.NodesEnd(); in++)
+        {
+            radius=h_factor*in->FastGetSolutionStepValue(NODAL_H);
+
+            work_point[0]=in->X();
+            work_point[1]=in->Y();
+            work_point[2]=in->Z();
+
+            n_points_in_radius = nodes_tree1.SearchInRadius(work_point, radius, res.begin(),res_distances.begin(), max_results);
+            if (n_points_in_radius>1)
+            {
+                if (in->FastGetSolutionStepValue(IS_BOUNDARY)==0.0 && in->FastGetSolutionStepValue(IS_STRUCTURE)==0.0)
+                {
+                    //look if we are already erasing any of the other nodes
+                    double erased_nodes = 0;
+                    for(PointIterator i=res.begin(); i!=res.begin() + n_points_in_radius ; i++)
+                       erased_nodes += in->Is(TO_ERASE);
+
+                    if( erased_nodes < 1) //we cancel the node if no other nodes are being erased
+                         in->Set(TO_ERASE,true);
+
+                }
+		//GLASS: WE DISTNIGUISH  THE INBLOW BY FLAG=1
+                else if ( (in)->FastGetSolutionStepValue(IS_STRUCTURE)!=1.0 && (in)->FastGetSolutionStepValue(FLAG_VARIABLE)!=1.0) //boundary nodes will be removed if they get REALLY close to another boundary node (0.2 * h_factor)
+                {
+                    //here we loop over the neighbouring nodes and if there are nodes
+                    //with IS_BOUNDARY=1 which are closer than 0.2*nodal_h from our we remove the node we are considering
+                    unsigned int k = 0;
+                    unsigned int counter = 0;
+                    for(PointIterator i=res.begin(); i!=res.begin() + n_points_in_radius ; i++)
+                    {
+                        if ( (*i)->FastGetSolutionStepValue(IS_BOUNDARY,1)==1.0 && res_distances[k] < 0.2*radius && res_distances[k] > 0.0 )
```

Comment:
Not in this pull request but in the future it would be great to change the use of variable IS_BOUNDARY to the flag:

```c++
if((*i)->Is(BOUNDARY)  && res_distances[k]< 0.2*radius && res_distances[k] > 0.0 )
```

It is faster and uses less memory.

Reply:
Will do so.  Thank you!